### PR TITLE
Be resilient to the document being shorter than when it started when getting completion description

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems?id=209299

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @rchande

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 1]
```diff
@@ -214,22 +216,28 @@ public static SupportedPlatformData GetSupportedPlatforms(CompletionItem item, W
...
-            if (item.Properties.TryGetValue("ContextPosition", out var text) && int.TryParse(text, out var number))
+            if (item.Properties.TryGetValue("ContextPosition", out var text) &&
+                int.TryParse(text, out var number))
             {
-                return number;
+                // We have no access to the editor at this layer.  So it's not 
+                // possible for us to map the original context position forward
+                // to the current position in the file.  So we need to cap the
+                // positoin to make sure it's within the bounds of the current
+                // text.
```
Comment (tmat) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: Is this the right fix? It seems to me that the problem is that CompletionItem is not associated with a snapshot of the document it was calculated for.

Comment (rchande) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: We dismiss the session if the caret goes left of the start of "word" boundary we choose, right? Can we just make context position be that left bound?

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: We're not really at a layer where we can associate with any good snapshot.  And associating with the Document itself would mean that completion items would keep your entire solution snapshot alive.

**Comment by dpoeschl** [comment]
Buddy tested. Looks good.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski

**Comment by CyrusNajmabadi** [comment]
Tagging @jasonmalinowski

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @Pilchie

File: src/Features/Core/Portable/Completion/CompletionService.cs [hunk 2]
```diff
@@ -169,5 +170,45 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
...
+            var document = item?.Document;
+            if (document == null || service == null)
+            {
+                return CompletionDescription.Empty;
+            }
+
+            return await service.GetDescriptionAsync(item.Document, item, cancellationToken).ConfigureAwait(false);
+        }
+
+        internal static async Task<CompletionList> GetCompletionsAsync(
```
Comment (Pilchie) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: Having a static method on `CompletionService` that requires an instance of a `CompletionService` is just wrong.

File: src/Features/Core/Portable/Completion/CompletionService.cs [hunk 3]
```diff
@@ -169,5 +170,45 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
...
+            CompletionService service, Document documentOpt, int caretPosition,
+            CompletionTrigger trigger = default, ImmutableHashSet<string> roles = null,
+            OptionSet options = null, CancellationToken cancellationToken = default)
+        {
+            if (documentOpt == null || service == null)
```
Comment (Pilchie) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: Why is it `Opt`, if you basically require it.?

File: src/Features/Core/Portable/Completion/CompletionService.cs (hunk 2 above)
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: Agreed. :smile:

Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: Why is this a static method at all? You already have a CompletionService in hand.

File: src/Features/Core/Portable/Completion/CompletionService.cs [hunk 4]
```diff
@@ -169,5 +170,45 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
...
+            {
+                return null;
+            }
+
+            var completions = await service.GetCompletionsAsync(
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: Why call this `completions` when it returns a `completionList`?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_ComputeModel.cs [hunk 5]
```diff
@@ -100,7 +100,8 @@ public void Do()
 
                         // TODO(cyrusn): We're calling into extensions, we need to make ourselves resilient
                         // to the extension crashing.
-                        var completionList = await GetCompletionListAsync(_completionService, _trigger, cancellationToken).ConfigureAwait(false);
+                        var completionList = await CompletionService.GetCompletionsAsync(
+                            _completionService, _documentOpt, _subjectBufferCaretPosition, _trigger, _roles, _options, cancellationToken).ConfigureAwait(false);
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_ComputeModel.cs]: Yeah, this call site looks really fishy.

File: src/Features/Core/Portable/Completion/CompletionItem.cs [hunk 6]
```diff
@@ -54,6 +54,14 @@ public sealed class CompletionItem : IComparable<CompletionItem>
         /// </summary>
         public CompletionItemRules Rules { get; }
 
+        /// <summary>
+        /// The <see cref="Document"/> that this <see cref="CompletionItem"/> was
+        /// created for.  Not available to clients.  Only used by the Completion
+        /// subsystem itself for things like being able to go back to the originating
+        /// Document when doing things like getting descriptions.
+        /// </summary>
+        internal Document Document { get; set; }
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionItem.cs]: Maybe not part of this PR, but this would potentially be useful to clients.

File: src/Features/Core/Portable/Completion/CompletionService.cs [hunk 7]
```diff
@@ -169,5 +170,45 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
 
             return bestItems.ToImmutableAndFree();
         }
+
+        internal static async Task<CompletionDescription> GetDescriptionAsync(
+            CompletionService service, CompletionItem item, CancellationToken cancellationToken = default)
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: This looks thoroughly broken to have a static method on a service take an instance of the service. That should be an instance method IMO.

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: So the problem as an instance method is that we already have the instance method.  and we'll conflict.

Comment (DustinCampbell) [inline src/Features/Core/Portable/Completion/CompletionService.cs]: Give it a different name?

**Comment by CyrusNajmabadi** [comment]
@Pilchie @DustinCampbell Have addressed your feedback.  Please take another look when you have a chance.
